### PR TITLE
[Chips] Adding Ripple support for Chips

### PR DESCRIPTION
**Pull request by yarneo**
In this PR I am adding initial support for Ripple in Chips as an opt-in behavior.
The ripple will be placed in the .m file and not exposed in the header file, similarly to the existing Ink.
Clients who wish to modify the Ripple's color for state can use the existing inkColorForState API, similarly to how Ripple was integrated in Buttons.

There are currently some issues in fully integrating the Selected state that Stateful Ripple provides into Chips. There have been bugs filed that block this effort at its current state: #7265 #7266 #7267 

As part of the effort for: #7268

### CODE REVIEW

File: components/Chips/src/MDCChipCollectionViewCell.m
Comment (romoore) [inline components/Chips/src/MDCChipCollectionViewCell.m:103]: Is it important that this method happen before the `super` call? If yes, please add a comment to that effect so maintainers don't move it about thinking the order is not important.

File: components/Chips/src/MDCChipView.m
Comment (romoore) [inline components/Chips/src/MDCChipView.m:269]: I believe you should be correcting the RippleView's frame to match the ChipView's bounds.  That should allow the autoresizing mask to work correctly.  If that's incorrect, please disable the autoresizing mask in the `init`.

File: components/Chips/src/MDCChipView.m
Comment (romoore) [inline components/Chips/src/MDCChipView.m]: This shouldn't be necessary of the frame is corrected in `setEnableRippleBehavior`.

File: components/Chips/src/MDCChipView.m
Comment (romoore) [inline components/Chips/src/MDCChipView.m]: Please use `self.rippleView` instead of accessing the iVar directly.  Since it's a property, the best practice is to use the getter whenever possible.

**Review by romoore — DISMISSED** [review]
Looks like this PR is missing unit/snapshot tests. Please add new tests for:

*  The new APIs added to MDCChipView and used by the collection view cell.
*  Snapshots for the new Ripple appearance.  Prefer to capture the Ripple effect in the middle of "spreading" as well as while held/highlighted.

File: components/Chips/src/MDCChipCollectionViewCell.m
Comment (yarneo) [reply components/Chips/src/MDCChipCollectionViewCell.m:103]: Yes it is and is mentioned in the called API header (MDCStatefulRippleView). I've added additional documentation if that wasn't clear.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m:269]: That is done in `layoutSubviews` and is not necessary here.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m]: It is necessary and is aligned with how inkView is set up, and the frame isn't corrected in the setter, but rather here.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m]: done

**Comment by yarneo** [comment]
Snapshot tests have been added in a previous PR for states in #7269 , is there any additional testing you seek?

**Comment by romoore** [comment]
@yarneo I would expect to see some new tests added for the new API/appearance. I'm not sure how you could have added snapshot tests for this code before the code landed.

File: components/Chips/src/MDCChipView.m
Comment (romoore) [reply components/Chips/src/MDCChipView.m:269]: Yes, I'm requesting that it be done once  when it changes (here) rather than explicitly overriding all the work done in `layoutSubviews` by the constraint system.

File: components/Chips/src/MDCChipView.m
Comment (romoore) [reply components/Chips/src/MDCChipView.m]: The frame could be fixed in the setter rather than doing the layout work twice here.  I think that's a good improvement over what was done for Ink.

File: components/Chips/src/MDCChipView.m
Comment (jverkoey) [reply components/Chips/src/MDCChipView.m:269]: +1 if we can rely on an autoresizing mask of width + height & frame update here to accomplish what we need.

File: components/Chips/src/MDCChipView.m
Comment (jverkoey) [inline components/Chips/src/MDCChipView.m]: Extract the 0.12 value to a static constant in this file please.

**Review by jverkoey — CHANGES_REQUESTED** [review]
Please add snapshot tests with the ripple behavior enabled.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m:269]: Can you please point me to where an auto resizing mask for width and height is set for the MDCChipView instance? I am unable to find it.

Nevermind, I assume you may be talking about the MDCRippleView's init. I'll give it a shot, although reluctant to change the consistency of how all the other subviews are set up. Also concerned this change of behavior may affect the results when a shapeGenerator is set.

My personal opinion is to stay consistent with the existing set up of the InkView and all the other subviews. So I see this improvement as risky.

Comment (yarneo) [reply components/Chips/src/MDCChipView.m:269]: Done.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m]: Done.

**Comment by yarneo** [comment]
Added snapshot tests and unit tests as requested.

File: components/Chips/tests/snapshot/MDCChipViewSnapshotTests.m
Comment (jverkoey) [inline components/Chips/tests/snapshot/MDCChipViewSnapshotTests.m:414]: Do any of these snapshot tests validate that the ripple is appearing as part of the state change? E.g. "with ripple enabled and the chip highlighted, the ripple appears".

File: components/Chips/src/MDCChipView.m
Comment (romoore) [inline components/Chips/src/MDCChipView.m:449]: Does the [Material Activated state](https://material.io/design/interaction/states.html#activated) need to be handled in this logic? Since a Chip can be either Selected (Filter Chips) or Activated (Choice Chips), when should `UIControlStateSelected` be mapped to the Activated Material (Ripple) state?

File: components/Chips/tests/snapshot/MDCChipViewSnapshotTests.m
Comment (romoore) [inline components/Chips/tests/snapshot/MDCChipViewSnapshotTests.m]: Nit: Remove extra space.

File: components/Chips/tests/unit/ChipViewRippleTests.m
Comment (romoore) [inline components/Chips/tests/unit/ChipViewRippleTests.m]: Tip: Why is this a "fake color"? It looks like a normal one.

File: components/Chips/tests/unit/ChipViewRippleTests.m
Comment (romoore) [inline components/Chips/tests/unit/ChipViewRippleTests.m:119]: Why doesn't Ripple support disabled states? [Disabled seems to be a Material state](https://material.io/design/interaction/states.html#disabled).  I would expect the Ripple view to handle any necessary overlay/color changes as a consequence of disabling the component.

**Review by romoore — COMMENTED** [review]
Overall seems good, just an open question about how Selected/Activated works with UIControlState and Ripple.

**Review by jverkoey — APPROVED** [review]
LGTM but please add a test for the ripple behavior next.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m:449]: As far as I'm aware, we don't have the "activated" state as part of the ChipView or ChipCollectionCell API, but rather we treat it identically to Selected. Once we do support it, we may probably need to add support for it in the implementation file and maybe also support it in Ripple's states.

File: components/Chips/tests/snapshot/MDCChipViewSnapshotTests.m
Comment (yarneo) [reply components/Chips/tests/snapshot/MDCChipViewSnapshotTests.m]: Done

File: components/Chips/tests/unit/ChipViewRippleTests.m
Comment (yarneo) [reply components/Chips/tests/unit/ChipViewRippleTests.m:119]: The disabled state does not have any Ripple or Overlay involved in it. However, we probably should allow clients if they wish to add a ripple in the disabled state. This could be added to the Ripple component once we have the state system fully understood by the team.

File: components/Chips/tests/unit/ChipViewRippleTests.m
Comment (yarneo) [reply components/Chips/tests/unit/ChipViewRippleTests.m]: Done.

File: components/Chips/src/MDCChipView.h
Comment (galiak11) [inline components/Chips/src/MDCChipView.h]: It seems this determines if to use the MDCRippleView or not. 
Isn't MDCInkView functionality determined by the ink API?

File: components/Chips/src/MDCChipView.m
Comment (galiak11) [inline components/Chips/src/MDCChipView.m:412]: I think the ink terminology is confusing and it might be a good opportunity to deprecate it together with the ink API.
For that purpose, creating a new [setRippleColor:forState:] to replace this ink API would allow us to deprecate this API later.

File: components/Chips/src/MDCChipView.m
Comment (galiak11) [reply components/Chips/src/MDCChipView.m:449]: Currently, Material theming of the `activated` and `selected` states is the same, but I'm not sure if we want to enforced that as a single theming in our components.

Also, a `draggable` and an `error` state are missing here.

File: components/Chips/src/MDCChipView.m
Comment (galiak11) [inline components/Chips/src/MDCChipView.m:443]: What about state combinations? How are these themed?
Specifically, some state combinations are required to maintain visual indication of state while the component is in other persistent states. These combinations must be themable to allow the visual feedback. These are:
selected/activated + highlighted
selected/activated + dragged
error + highlighted

File: components/Chips/src/MDCChipView.m
Comment (galiak11) [inline components/Chips/src/MDCChipView.m:450]: Do we ever get to this condition? If either of these are true, a previous case will take precedence.

File: components/Chips/src/MDCChipView.h
Comment (galiak11) [inline components/Chips/src/MDCChipView.h:119]: When adding custom states - I recommend adding getters/setters for them in the API so they can be tested or used in examples.

File: components/Chips/src/MDCChipView.h
Comment (yarneo) [reply components/Chips/src/MDCChipView.h]: I don't understand your comment. This property decides if to use MDCInkView or MDCStatefulRippleView. If you need better clarification of the documentation please provide suggestions.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m:412]: That is up to your state work to decide the steps forward here. The task at hand here is to add Ripple as an opt-in behavior to Chips. It's not up to me to decide the API exposed for states and I don't want to add additional API that may be deprecated in the near future.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m:443]: Highlighted + Selected is supported. Activated == Selected currently in Chips so to add the Activated state is out of scope here. Also again, error and what not is undefined for Ripple as it is something I would need input from you to decide what the spec wants the Ripple to look like for that state. I recall you saying "Error doesn't need Ripple"

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m:450]: We do. And no previous cases don't take precedence, both states occurring at the same time creates a different integer.

File: components/Chips/src/MDCChipView.h
Comment (yarneo) [reply components/Chips/src/MDCChipView.h:119]: Adding new API for new states is out of scope for this PR.

File: components/Chips/src/MDCChipView.m
Comment (yarneo) [reply components/Chips/src/MDCChipView.m:449]: Please provide guidance as part of your states work regarding future work on the component to provide new states.

File: components/Chips/src/MDCChipView.h
Comment (galiak11) [reply components/Chips/src/MDCChipView.h]: I was referring to "if an @c MDCChipView should use the @c MDCInkView behavior or not". 
Suggestion: "if an @c MDCChipView should use the @c MDCRippleView behavior or not"

File: components/Chips/src/MDCChipView.m
Comment (galiak11) [reply components/Chips/src/MDCChipView.m:443]: I think only disabled doesn't require an overlay. All states that require an overlay will probably need Ripple. I doesn't make sense to create a competing alternative solution. 
I'm fine with creating a followup ticket/pr for these issues so we can keep the scope of this PR small.

File: components/Chips/src/MDCChipView.m
Comment (galiak11) [reply components/Chips/src/MDCChipView.m:449]: Sounds good. We can add these in a followup PR.

**Review by galiak11 — APPROVED** [review]

**Comment by yarneo** [comment]
Thanks Galia, updated the doc to the header file based on your feedback.

**Comment by yarneo** [comment]
I've opened https://github.com/material-components/material-components-ios/issues/7295 as a follow up to this PR to explore how we can better test the Ink/Ripple in snapshot tests for the ChipView, as it seems to be unable to present the Ink/Ripple properly despite the Ink/Ripple being there.

**Pull request merged**
